File: shared/kmp/src/commonMain/kotlin/lk/mageride/shared/data/models/PositionSample.kt
package lk.mageride.shared.data.models

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

/**
 * The canonical position event — one GNSS fix from one vehicle.
 *
 * This is the single shape that travels the whole telemetry plane: the driver app and every
 * hardware tracker publish it to `veh/{vehicleId}/pos/live`, `position-processor-svc` (C039)
 * republishes it onto `telemetry.normalized`, and `persistence-writer-svc` (C040) COPYs it into
 * the `telemetry.positions` hypertable. CBOR on the wire, JSON-equivalent here
 * (`backend/contracts/realtime/mqtt-topics.md` §2.1, D6' §2.2).
 *
 * **Spec divergence — ADD Appendix A.** The C012 deliverable anchors this at ADD "Appendix A —
 * Position Event Schema (Canonical)", which prints an older, looser shape: `ts` as epoch millis,
 * `source` as a free string (`"mobile_gps | hardware_gt06 | hardware_st901"`), plus `altitude`,
 * `tripSessionId` and `traceId`, and **no `seq`**. Three later sources agree against it —
 * `mqtt-topics.md` §2.1 (the machine-checkable contract), D6' §2.2, and the landed
 * `telemetry.positions` DDL (C006), whose columns are `sample_ts`/`received_ts`/`seq`/`speed_mps`/
 * `heading_deg`/`accuracy_m`/`hdop`/`sat_count`/`source SMALLINT CHECK (source BETWEEN 0 AND 4)`.
 * Modelling Appendix A would produce a DTO that cannot be written to its own sink and that omits
 * `seq`, which is the replay dedupe key R-17/T-05 exists for. **The later, runnable shape wins**;
 * a micro-change-set against ADD Appendix A is recorded in the C012 handoff.
 *
 * `seq` is monotonic per vehicle and is layer 3 of the replay dedupe
 * (`ux_positions_vehicle_seq (vehicle_id, seq, sample_ts)`). A tracker that reconnects after
 * buffering bursts to `pos/replay` carrying the sequence numbers and GNSS timestamps it captured,
 * so an exact duplicate is rejected by the database and a near-duplicate by the Redis
 * `veh:seq:{vehicleId}` watermark.
 *
 * C017 owns the MQTT client, the adaptive publish cadence and the CBOR codec that carries this;
 * C012 owns only the shape.
 *
 * @property vehicleId The publishing vehicle. EMQX binds it from the device credential, so a
 *   device physically cannot publish under another vehicle's topic.
 * @property sampleTs GNSS capture instant — **not** the receive time. The hypertable's time
 *   dimension.
 * @property receivedTs When the platform saw it. `sampleTs` − `receivedTs` is the replay lag.
 * @property seq Monotonic per vehicle, `>= 0`. The replay dedupe key (R-17, T-05).
 * @property lat Degrees, −90…90. Range-checked by `ck_positions_lat`.
 * @property lng Degrees, −180…180. Range-checked by `ck_positions_lng`.
 * @property speedMps Ground speed in metres per second.
 * @property headingDeg Course over ground, 0…359.
 * @property accuracyM Horizontal accuracy in metres.
 * @property hdop Horizontal dilution of precision, as the receiver reports it.
 * @property satCount Satellites used in the fix.
 * @property source Which stack produced the fix.
 * @property mode The vehicle's operating mode at capture time.
 * @property vehicleType The vehicle's type, denormalised so a consumer needs no registry lookup.
 * @property fleetId Denormalised at write time for fleet-scoped reads and RLS; `null` when the
 *   vehicle is not fleet-owned. A vehicle that changes fleet keeps its old rows under the old
 *   fleet, which is what an audit trail should do (C006 decision 8).
 * @property tripId The Mode A/B tracking session this belongs to; absent for Mode C.
 */
@Serializable
public data class PositionSample(
    val vehicleId: Ulid,
    val sampleTs: Timestamp,
    val receivedTs: Timestamp? = null,
    val seq: Long,
    val lat: Double,
    val lng: Double,
    val speedMps: Double? = null,
    val headingDeg: Int? = null,
    val accuracyM: Double? = null,
    val hdop: Double? = null,
    val satCount: Int? = null,
    val source: PositionSource,
    val mode: ServiceMode? = null,
    val vehicleType: VehicleType? = null,
    val fleetId: Ulid? = null,
    val tripId: Ulid? = null,
) {
    /** The fix as a plain coordinate. */
    public val point: GeoPoint get() = GeoPoint(lat = lat, lng = lng)
}

/**
 * Which stack produced a [PositionSample] (`telemetry.positions.source`, `ck_positions_source`).
 *
 * **Encoded as a small integer, not a name** — the column is a `SMALLINT` constrained to `0…4`
 * and the CBOR payload carries the same number, so the enum serialises through
 * [PositionSourceSerializer] rather than by `@SerialName`.
 *
 * @property code The wire value, `0…4`.
 */
@Serializable(with = PositionSourceSerializer::class)
public enum class PositionSource(public val code: Int) {
    /** The driver app's own GPS. */
    MOBILE(0),

    /** Concox GT06 / GT06N family, including TK103 and ST-901, via `adapter-gt06`. */
    GT06(1),

    /** JT/T 808 trackers via `adapter-jt808`. */
    JT808(2),

    /** H02 / H02X ASCII bus trackers via `adapter-h02`. */
    H02(3),

    /** Teltonika / Queclink firmware speaking NMEA over native MQTT — no adapter in the path. */
    NMEA_MQTT(4),
    ;

    /** Whether the fix came from a hardware tracker rather than the driver's handset (T-11). */
    public val isHardware: Boolean get() = this != MOBILE

    public companion object {
        private val BY_CODE: Map<Int, PositionSource> = entries.associateBy { it.code }

        /** Resolves a wire code, or `null` when it is outside the `0…4` CHECK domain. */
        public fun fromCode(code: Int): PositionSource? = BY_CODE[code]
    }
}

/** Encodes [PositionSource] as the `0…4` integer the topic contract and the DDL both use. */
public object PositionSourceSerializer : KSerializer<PositionSource> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("lk.mageride.PositionSource", PrimitiveKind.INT)

    override fun serialize(encoder: Encoder, value: PositionSource) {
        encoder.encodeInt(value.code)
    }

    override fun deserialize(decoder: Decoder): PositionSource {
        val code = decoder.decodeInt()
        return PositionSource.fromCode(code)
            ?: throw SerializationException("Unknown PositionSource code: $code")
    }
}

File: backend/src/MageRide.Shared/Telemetry/PositionSample.cs
using MageRide.Shared.Primitives;

namespace MageRide.Shared.Telemetry;

/// <summary>
/// Which stack produced a fix (<c>telemetry.positions.source</c>, <c>ck_positions_source</c>).
/// </summary>
/// <remarks>
/// Encoded as the small integer the column's CHECK constrains to <c>0…4</c> and the CBOR payload
/// carries — not as a name. The KMP module's <c>PositionSource</c> serialises the same way.
/// </remarks>
public enum PositionSource
{
    /// <summary>The driver app's own GPS.</summary>
    Mobile = 0,

    /// <summary>Concox GT06 / GT06N family, including TK103 and ST-901, via <c>adapter-gt06</c>.</summary>
    Gt06 = 1,

    /// <summary>JT/T 808 trackers via <c>adapter-jt808</c>.</summary>
    Jt808 = 2,

    /// <summary>H02 / H02X ASCII bus trackers via <c>adapter-h02</c>.</summary>
    H02 = 3,

    /// <summary>Teltonika / Queclink firmware speaking NMEA over native MQTT — no adapter.</summary>
    NmeaMqtt = 4,
}

/// <summary>
/// The canonical position event — one GNSS fix from one vehicle
/// (<c>backend/contracts/realtime/mqtt-topics.md</c> §2.1, D6' §2.2).
/// </summary>
/// <remarks>
/// <para>
/// This is the single shape that travels the whole telemetry plane: the driver app and every
/// hardware tracker publish it to <c>veh/{vehicleId}/pos/live</c>, position-processor-svc
/// republishes it onto <c>telemetry.normalized</c>, and persistence-writer-svc (C040) will COPY it
/// into the <c>telemetry.positions</c> hypertable. It is the .NET half of the KMP module's
/// <c>lk.mageride.shared.data.models.PositionSample</c> (C012) and the two must stay identical —
/// the app encodes what this decodes.
/// </para>
/// <para>
/// <b>Spec divergence — ADD Appendix A.</b> Appendix A prints an older, looser shape: <c>ts</c> as
/// epoch millis, <c>source</c> as a free string, plus <c>altitude</c>/<c>tripSessionId</c>/
/// <c>traceId</c>, and <b>no <c>seq</c></b>. Three later sources agree against it —
/// <c>mqtt-topics.md</c> §2.1, D6' §2.2, and the landed <c>telemetry.positions</c> DDL (C006), whose
/// columns are exactly the members below. Modelling Appendix A would produce a DTO that cannot be
/// written to its own sink and that omits <c>seq</c>, the replay dedupe key R-17/T-05 exists for.
/// The later, runnable shape wins; the micro-change-set is already recorded in the C012 handoff.
/// </para>
/// <para>
/// Every optional member is nullable because a cheap tracker reports few of them and the CBOR
/// encoder omits defaults. <c>lat</c>, <c>lng</c>, <c>seq</c> and <c>sampleTs</c> are not optional:
/// a sample without them is not a position.
/// </para>
/// </remarks>
/// <param name="VehicleId">The publishing vehicle. EMQX binds it from the device credential, so a
/// device physically cannot publish under another vehicle's topic.</param>
/// <param name="SampleTs">GNSS capture instant — <b>not</b> the receive time. The hypertable's time
/// dimension.</param>
/// <param name="Seq">Monotonic per vehicle, <c>&gt;= 0</c>. The replay dedupe key (R-17, T-05).</param>
/// <param name="Lat">Degrees, −90…90. Range-checked by <c>ck_positions_lat</c>.</param>
/// <param name="Lng">Degrees, −180…180. Range-checked by <c>ck_positions_lng</c>.</param>
/// <param name="Source">Which stack produced the fix.</param>
/// <param name="ReceivedTs">When the platform saw it. The gap from <paramref name="SampleTs"/> is
/// the replay lag.</param>
/// <param name="SpeedMps">Ground speed in metres per second.</param>
/// <param name="HeadingDeg">Course over ground, 0…359.</param>
/// <param name="AccuracyM">Horizontal accuracy in metres.</param>
/// <param name="Hdop">Horizontal dilution of precision, as the receiver reports it.</param>
/// <param name="SatCount">Satellites used in the fix.</param>
/// <param name="Mode">The vehicle's operating mode at capture time — <c>A</c>, <c>B</c> or
/// <c>C</c>.</param>
/// <param name="VehicleType">Canonical vehicle type, denormalised so a consumer needs no registry
/// lookup.</param>
/// <param name="FleetId">Denormalised for fleet-scoped reads and RLS; null when the vehicle is not
/// fleet-owned.</param>
/// <param name="TripId">The Mode A/B tracking session this belongs to; absent for Mode C.</param>
public sealed record PositionSample(
    Guid VehicleId,
    DateTimeOffset SampleTs,
    long Seq,
    double Lat,
    double Lng,
    PositionSource Source,
    DateTimeOffset? ReceivedTs = null,
    double? SpeedMps = null,
    int? HeadingDeg = null,
    double? AccuracyM = null,
    double? Hdop = null,
    int? SatCount = null,
    string? Mode = null,
    string? VehicleType = null,
    Guid? FleetId = null,
    Guid? TripId = null)
{
    /// <summary>The fix as a plain coordinate.</summary>
    public GeoPoint Point => new(Lat, Lng);

    /// <summary>
    /// Whether the sample is inside the CHECK domains <c>telemetry.positions</c> enforces.
    /// </summary>
    /// <remarks>
    /// <c>mqtt-topics.md</c> §2.1: "a cheap tracker reporting 0/999 degrees is a bug C039 must filter
    /// before the batch arrives — the constraints make it loud instead of silently poisoning a
    /// rollup." This is that filter's cheap half; the anti-spoof work (speed plausibility, jump
    /// detection, the Kalman pass) is C039/C040 and is deliberately not here.
    /// </remarks>
    public bool IsWellFormed =>
        Seq >= 0
        && !double.IsNaN(Lat) && Lat is >= -90 and <= 90
        && !double.IsNaN(Lng) && Lng is >= -180 and <= 180
        && Enum.IsDefined(Source)
        && VehicleId != Guid.Empty;
}
